### ISSUE TITLE
Revist Immutability

### ISSUE THREAD

**Issue by nicholaides** [picwell/schematics-extensions#15]
As @dmcclory mentioned elsewhere:

> We should revisit the immutable model. It'd be nice if there was a way to update an attribute and get a dup'd copy.

> ```
> x.schematics_attribute # =>100
> 
> y = x.set('schematics_attribute', 200)
> 
> y.schematics_attribute # => 200
> x.schematics_attribute # =>100
> y.other_field == x.other_field # => True
> ```

Another idea would be to use a context manager to break immutability while initializing:

    with x.temporarily_allow_mutations:
        x.some_attribute = 34

**Comment by csojinb-picwell** [comment]
I'd much prefer returning a new object with the new attribute. Maintains the desired property that the input model won't be changed, which could cause problems if it needed to be used elsewhere.

**Comment by csojinb-picwell** [comment]
@dmcclory @nicholaides I'm running back through these issues, and I think I misread Dan's code example the first time around. It looks like the `x.set` _is_ creating a new object, with a new value for `schematics_attribute`, yes? If so, I'm on board with the concept, though maybe we could find a different name than `set`, as I think that might create the expectation that `x` is changing.

**Comment by dmcclory** [comment]
It'd be good if it was a simple word. `change`, `replace`, `update`, as opposed to `set_on_new_copy`.

 But maybe the best thing would be to define `__setitem__(...)` so that it returns a new copy. Then it'd be possible to set properties with `model['key'] = `  (and possibly `model.key = `). That'd look the most like the syntax used on a normal object, when we destructively setting a property.

**Comment by csojinb-picwell** [comment]
Yeah, I agree on the simple word. I think I like `replace` best of the options suggested.

I was actually thinking it would be good for the syntax to be a little different so that people don't get confused between things that do destructively set a property and things that don't? For example, pandas DataFrames have a non-mutating `append` method (returns a new DataFrame), but python `list.append` mutates the list, and I think I've spent at least a month of my life resolving in how I used `append` on something.

That said, I can also see the argument for familiar syntax
